packages/inferno/src/DOM/unmounting.ts

```diff
@@ -35,53 +35,82 @@ export function unmount(vNode) {
     } else if (childFlags === ChildFlags.HasVNodeChildren) {
-      unmount(children as VNode);
+      unmount(children as VNode, animations);
     }
   } else if (children) {
     if (flags & VNodeFlags.ComponentClass) {
       if (isFunction(children.componentWillUnmount)) {
+        // TODO: Possible entrypoint
         children.componentWillUnmount();
       }
+
+      // If we have a willDisappear on this component, block children
+      let childAnimations = animations;
+      if (isFunction(children.willDisappear)) {
+        childAnimations = new AnimationQueues();
+        addDisappearAnimationHook(animations, children, children.$LI.dom);
+      }
+
       unmountRef(vNode.ref);
       children.$UN = true;
-      unmount(children.$LI);
+      unmount(children.$LI, childAnimations);
     } else if (flags & VNodeFlags.ComponentFunction) {
       ref = vNode.ref;
 
       if (!isNullOrUndef(ref) && isFunction(ref.onComponentWillUnmount)) {
+        // TODO: Possible entrypoint
         ref.onComponentWillUnmount(findDOMfromVNode(vNode, true) as Element, vNode.props || EMPTY_OBJ);
       }
 
-      unmount(children);
+      unmount(children, animations);
     } else if (flags & VNodeFlags.Portal) {
-      remove(children as VNode, vNode.ref);
+      // TODO: Possible entrypoint
+      remove(children as VNode, vNode.ref, animations);
     } else if (flags & VNodeFlags.Fragment) {
       if (vNode.childFlags & ChildFlags.MultipleChildren) {
-        unmountAllChildren(children);
+        // TODO: Possible entrypoint
+        unmountAllChildren(children, animations);
       }
     }
   }
 }
 
-export function unmountAllChildren(children: VNode[]) {
+export function unmountAllChildren(children: VNode[], animations: AnimationQueues) {
   for (let i = 0, len = children.length; i < len; ++i) {
-    unmount(children[i]);
+    unmount(children[i], animations);
   }
 }
 
-export function clearDOM(dom) {
-  // Optimization for clearing dom
-  dom.textContent = '';
+export function clearDOM(dom, animations: AnimationQueues) {
+  if (animations.willDisappear.length > 0) {
+    // Wait until animations are finished before removing actual dom nodes
+    // Be aware that the element could be removed by a later operation
+    callAllAnimationHooks(animations.willDisappear, () => { if (dom) dom.textContent = '' });
+  }
+  else {
+    // Optimization for clearing dom
+    dom.textContent = '';
+  }
 }
 
-export function removeAllChildren(dom: Element, vNode: VNode, children) {
-  unmountAllChildren(children);
+export function removeAllChildren(dom: Element, vNode: VNode, children, animations: AnimationQueues) {
+  unmountAllChildren(children, animations);
 
   if (vNode.flags & VNodeFlags.Fragment) {
-    removeVNodeDOM(vNode, dom);
+    removeVNodeDOM(vNode, dom, animations);
   } else {
-    clearDOM(dom);
+    clearDOM(dom, animations);
   }
 }
+
+function addDisappearAnimationHook(animations: AnimationQueues, instance, dom: Element) {
```

Comment:
same comment about hoisting window !== 'undefined'

Reply:
Will fix

Reply:
FIXED in e2c8e0b08